ddl/ddl.go

```diff
@@ -642,45 +646,106 @@ func (d *ddl) prepareWorkers4legacyDDL() {
 
 // Start implements DDL.Start interface.
 func (d *ddl) Start(ctxPool *pools.ResourcePool) error {
-	logutil.BgLogger().Info("[ddl] start DDL", zap.String("ID", d.uuid), zap.Bool("runWorker", RunWorker))
+	var err error
+	logutil.BgLogger().Info("[ddl] start DDL", zap.String("ID", d.uuid), zap.Bool("runWorker", config.GetGlobalConfig().Instance.TiDBEnableDDL.Load()))
 
 	d.wg.Run(d.limitDDLJobs)
 	d.sessPool = newSessionPool(ctxPool, d.store)
-	// If RunWorker is true, we need campaign owner and do DDL job.
+
+	// If tidb_enable_ddl is true, we need campaign owner and do DDL job.
 	// Otherwise, we needn't do that.
-	if RunWorker {
-		d.ownerManager.SetBeOwnerHook(func() {
-			var err error
-			d.ddlSeqNumMu.seqNum, err = d.GetNextDDLSeqNum()
-			if err != nil {
-				logutil.BgLogger().Error("error when getting the ddl history count", zap.Error(err))
-			}
-		})
+	if config.GetGlobalConfig().Instance.TiDBEnableDDL.Load() {
+		if err = d.EnableDDL(); err != nil {
+			return err
+		}
+	}
+
+	variable.RegisterStatistics(d)
+
+	metrics.DDLCounter.WithLabelValues(metrics.CreateDDLInstance).Inc()
+
+	// Start some background routine to manage TiFlash replica.
+	d.wg.Run(d.PollTiFlashRoutine)
 
-		err := d.ownerManager.CampaignOwner()
+	return nil
+}
+
+// EnableDDL enable this node to execute ddl.
+// Since ownerManager.CampaignOwner will start a new goroutine to run ownerManager.campaignLoop,
+// we should make sure that before invoking EnableDDL(), ddl is DISABLE.
+func (d *ddl) EnableDDL() error {
+	d.ownerManager.SetBeOwnerHook(func() {
+		var err error
+		d.ddlSeqNumMu.seqNum, err = d.GetNextDDLSeqNum()
 		if err != nil {
-			return errors.Trace(err)
+			logutil.BgLogger().Error("error when getting the ddl history count", zap.Error(err))
 		}
+	})
 
-		d.delRangeMgr = d.newDeleteRangeManager(ctxPool == nil)
+	err := d.ownerManager.CampaignOwner()
+	if err != nil {
+		return errors.Trace(err)
+	}
 
-		d.prepareWorkers4ConcurrencyDDL()
-		d.prepareWorkers4legacyDDL()
+	d.delRangeMgr = d.newDeleteRangeManager(d.sessPool.resPool == nil)
 
-		go d.schemaSyncer.StartCleanWork()
-		if config.TableLockEnabled() {
-			d.wg.Add(1)
-			go d.startCleanDeadTableLock()
-		}
-		metrics.DDLCounter.WithLabelValues(metrics.StartCleanWork).Inc()
+	d.prepareWorkers4ConcurrencyDDL()
+	d.prepareWorkers4legacyDDL()
+
+	go d.schemaSyncer.StartCleanWork()
+	if config.TableLockEnabled() {
+		d.wg.Add(1)
+		go d.startCleanDeadTableLock()
 	}
+	metrics.DDLCounter.WithLabelValues(metrics.StartCleanWork).Inc()
+	return errors.Trace(err)
+}
 
-	variable.RegisterStatistics(d)
+// DisableDDL disable this node to execute ddl.
+// We should make sure that before invoking DisableDDL(), ddl is ENABLE.
+func (d *ddl) DisableDDL() error {
```

Comment:
`DisableDDL()` seems to be non-revertible. I test it with the following steps:

1. Start two TiDB server, say `tidb-a`(DDL owner) and `tidb-b`.
2. Execute on `tidb-a`: `set @@tidb_enable_ddl = false;`
3. Stop `tidb-a`.
4. Execute on `tidb-a`: `set @@tidb_enable_ddl = true;`
5. Execute on `tidb-a`: `create table t (a int);`

The last step hangs.